contracts/distribution/dao-rewards-distributor/src/contract.rs

```diff
@@ -0,0 +1,683 @@
+
+    execute_fund(deps, env, sender, reward_denom_config, wrapper.amount)
+}
+
+fn execute_fund_native(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+) -> Result<Response, ContractError> {
+    let fund_coin = one_coin(&info).map_err(|_| ContractError::InvalidFunds {})?;
+
+    let reward_denom_config = REWARD_DENOM_CONFIGS.load(deps.storage, fund_coin.denom.clone())?;
+
+    execute_fund(
+        deps,
+        env,
+        info.sender,
+        reward_denom_config,
+        fund_coin.amount,
+    )
+}
+
+fn execute_fund(
+    mut deps: DepsMut,
+    env: Env,
+    sender: Addr,
+    mut denom_reward_config: DenomRewardConfig,
+    amount: Uint128,
+) -> Result<Response, ContractError> {
+    // Ensure that the sender is the owner
+    cw_ownable::assert_owner(deps.storage, &sender)?;
+
+    let denom_str = denom_reward_config.to_str_denom();
+
+    // first we update the existing rewards (if any)
+    update_rewards(&mut deps, &env, &sender, vec![denom_str.to_string()])?;
+
+    // we derive the period for which the rewards are funded
+    // by looking at the existing reward emission config and the funded amount
+    let funded_period_duration = denom_reward_config
+        .reward_emission_config
+        .get_funded_period_duration(amount)?;
+    let funded_period_units = get_duration_units(&funded_period_duration);
+
+    denom_reward_config = denom_reward_config
+        .bump_funding_date(&env.block)
+        .bump_last_update(&env.block);
+
+    denom_reward_config.distribution_expiration = match denom_reward_config.distribution_expiration
+    {
+        // if this is the first funding of the denom, the new expiration is the funded period duration
+        // from the current block
+        Expiration::Never {} => funded_period_duration.after(&env.block),
+        // otherwise we add the duration units to the existing expiration
+        Expiration::AtHeight(h) => {
+            if h <= env.block.height {
+                // expiration is the funded duration after current block
+                Expiration::AtHeight(env.block.height + funded_period_units)
+            } else {
+                // if the previous expiration had not yet expired, we extend
+                // the current rewards period by the newly funded duration
+                Expiration::AtHeight(h + funded_period_units)
+            }
+        }
+        Expiration::AtTime(t) => {
+            if t.seconds() <= env.block.time.seconds() {
+                // expiration is the funded duration after current block time
+                Expiration::AtTime(t.plus_seconds(funded_period_units))
+            } else {
+                // if the previous expiration had not yet expired, we extend
+                // the current rewards period by the newly funded duration
+                Expiration::AtTime(env.block.time.plus_seconds(funded_period_units))
+            }
+        }
```

Comment:
```suggestion
        Expiration::AtTime(t) => {
            if t.seconds() <= env.block.time.seconds() {
                // expiration is the funded duration after current block time
                Expiration::AtTime(env.block.time.plus_seconds(funded_period_units))
            } else {
                // if the previous expiration had not yet expired, we extend
                // the current rewards period by the newly funded duration
                Expiration::AtTime(t.plus_seconds(funded_period_units))
            }
        }
```

i think you flipped the cases. the first one should use the current block time, and the second one should extend the current period

Reply:
i made the fix but pls verify that i'm right